File: ios/SuperSecureChat/SscNativeBridge.swift
import WebKit

enum SscNativeBridge {
    static func handle(method: String, callbackId: String, payload: String, webView: WKWebView) {
        do {
            let args = try JSONSerialization.jsonObject(with: Data(payload.utf8)) as? [String: Any] ?? [:]
            let result = try SscCryptoService.shared.dispatch(method: method, args: args)
            deliver(callbackId: callbackId, ok: true, payload: result, webView: webView)
        } catch {
            deliver(callbackId: callbackId, ok: false, payload: error.localizedDescription, webView: webView)
        }
    }

    private static func deliver(callbackId: String, ok: Bool, payload: Any, webView: WKWebView) {
        let encoded: String
        if let dict = payload as? [String: Any],
           let data = try? JSONSerialization.data(withJSONObject: dict),
           let text = String(data: data, encoding: .utf8) {
            encoded = text
        } else if let text = payload as? String {
            encoded = "\"\(text.replacingOccurrences(of: "\\", with: "\\\\").replacingOccurrences(of: "\"", with: "\\\""))\""
        } else {
            encoded = "\"ssc_crypto_error\""
        }
        let script = "window.__sscBridge && window.__sscBridge._callback('\(callbackId)', \(ok ? "true" : "false"), \(encoded));"
        DispatchQueue.main.async {
            webView.evaluateJavaScript(script, completionHandler: nil)
        }
    }
}

File: android/app/src/main/java/com/supersecurechat/app/SscNativeBridge.kt
package com.supersecurechat.app

import android.app.Activity
import android.webkit.JavascriptInterface
import android.webkit.WebView
import org.json.JSONObject
import java.util.concurrent.Executors

/**
 * Async libsignal bridge for WebView — mirrors Electron preload `window.sscCrypto`.
 */
class SscNativeBridge(
    private val activity: Activity,
    private val webView: WebView,
    filesDir: java.io.File,
) {
    private val executor = Executors.newSingleThreadExecutor()

    init {
        SscCryptoService.bind(filesDir)
    }

    @JavascriptInterface
    fun fetchApi(url: String, method: String, headersJson: String, body: String, callbackId: String) {
        executor.execute {
            try {
                val apiBase = BuildConfig.SSC_API_URL
                val result = SscApiBridge.fetch(url, method, headersJson, body.ifBlank { null }, apiBase)
                deliver(callbackId, result.getBoolean("ok"), result)
            } catch (error: Throwable) {
                deliver(callbackId, false, error.message ?: "ssc_api_error")
            }
        }
    }

    @JavascriptInterface
    fun invoke(method: String, callbackId: String, argsJson: String) {
        executor.execute {
            try {
                val args = if (argsJson.isBlank()) JSONObject() else JSONObject(argsJson)
                val result = when {
                    method.startsWith("translate") -> SscTranslateService.dispatch(method, args)
                    else -> SscCryptoService.dispatch(method, args)
                }
                deliver(callbackId, true, result)
            } catch (error: Throwable) {
                deliver(callbackId, false, error.message ?: "ssc_native_error")
            }
        }
    }

    private fun deliver(callbackId: String, ok: Boolean, payload: Any) {
        val encoded = when (payload) {
            is String -> JSONObject.quote(payload)
            is JSONObject -> payload.toString()
            else -> JSONObject.quote(payload.toString())
        }
        val js = "window.__sscBridge._callback(${JSONObject.quote(callbackId)},$ok,$encoded);"
        activity.runOnUiThread {
            webView.evaluateJavascript(js, null)
        }
    }
}